irmin-pack: introduce shallow GC API changes, V0 GC compatibility

Firobe opened the pull request:
This supersedes (https://github.com/mirage/irmin/pull/2168) and adds V0 compatibility in the GC reachability algorithm as well as initial changes in the GC API.

The GC finalization for archival mode will be implemented in a further PR.

metanivek commented:
@Firobe are you looking for feedback?

Firobe commented:
Sure, feel free to roast what's already pushed :)

metanivek commented on src/irmin-pack/unix/store.ml:
```diff
@@ -392,7 +394,7 @@ module Maker (Config : Conf.S) = struct
         end
 
-        let is_split_allowed = Gc.is_allowed
+        let is_split_allowed = Gc.is_allowed ~action:`Delete
```
This isn't quite right since split should be allowed for stores where GC is allowed when archiving.

To achieve this, you could add something like a `gc_action` argument to pass through to `Gc.is_allowed`. You would also need to add this argument to the `split` function in order to check the appropriate `is_split_allowed`. At first pass, it seems a bit unfortunate to dirty the split api with gc action concerns, but it has the benefit of more clearly communicating the intrinsic connection between split and gc.

Firobe replied:
I was worrying about dirtying the `split` API indeed, but this is probably necessary for now indeed.

Firobe replied:
Though this makes me think: does it make sense for the same store to be gc'ed in archival and deletion mode at different times?

Would it be fair to instead choose the GC action at store opening (or even creation), and use that choice implicitly internally instead of re-exposing the choice on each API call? (at the risk of the user mixing flags by error).

metanivek replied:
Aha, you're right that it doesn't make sense to mix GC calls of archive and delete. I really like your idea of embedding this as a configuration choice... and we already have that in the `lower_root`! We can decide which type of GC to use based on this being set or not:

```ocaml
let action = match lower_root with
| Some _ -> `Archive
| None -> `Delete
```
So we would never have to expose this to the user and thus keep our API nice and tidy. :nail_care: 

In hindsight, keeping the action as a flag on the GC API is really a remnant of the original purpose which was to carry a volume identifier and thus trigger potential creation of a new volume. With the removal of the volume identifier and signalling of volume creation, I think we can go the next step and remove the action entirely!

What do you think?

Firobe replied:
I really like this! We just need to make really clear the implications of setting `lower_root` at store opening :)

metanivek replied:
Nice! Yeah I agree. We should update the docs for the config value to make this clear.

Also, once we have the entire mechanism in place, we should either update the [existing gc example](https://github.com/mirage/irmin/blob/main/examples/irmin-pack/gc.ml) or create a new one that uses the lower layer.

metanivek commented on src/irmin-pack/unix/gc.ml:
```diff
@@ -267,6 +272,16 @@ module Make (Args : Gc_args.S) = struct
                 let () = Lwt.wakeup_later t.resolver (Ok stats) in
                 Ok (`Finalised stats)
+            | `Success, `Archive, Ok _ ->
+                (* TODO actually archive garbage into lower volume *)
```
Yip, to be clear, this is A, B, and C in https://hackmd.io/jRgzxlL4Qry6SD3k4A_Z2A#Updates-to-GC-API-and-process

metanivek reviewed (commented):
Left one comment about checking split availability that needs addressing, but this is looking good!

Firobe commented:
Following our discussion with @metanivek, removed the various flags in the API and instead now we choose the behavior of GC finalisation depending on the presence of a lower layer. Also added a `behavior` function for the user to check that.

metanivek commented on CHANGES.md:
```diff
@@ -8,15 +8,21 @@
   - Add `is_split_allowed` to check if a store allows split. (#2175, @metanivek)
   - Add `add_volume` to allow creating new empty volume in lower layer. (#2188,
     @metanivek)
+  - Add a `behavior` function to the GC to check wether the GC will archive or
```
I've usually seen standard British English spellings (eg, finalise vs finalize) in the project, but I don't know if this is an intentional style choice. @samoht any input on this?

Firobe replied:
I actually wondered when choosing the name :p A quick grep shows a mixed usage but indeed more occurrences of `behaviour`

metanivek replied:
Hmm, I don't have a strong opinion on this, so we can go with whichever you prefer. :smile:

Firobe replied:
Went with `behaviour` :p

samoht replied:
The first version of Irmin was written at Cambridge - it inherited the British English spelling ;p 

I don't have a strong opinion, but I try to keep this consistent with the rest of the codebase (so I tend to continue to favour British English in comments).
